controllers/volumehandler/volumehandler.go

```diff
@@ -452,10 +490,115 @@ func (vh *VolumeHandler) ensureSnapshot(ctx context.Context, log logr.Logger,
 	// status.readyToUse either is not set by the driver at this point (even though
 	// status.BoundVolumeSnapshotContentName is set), or readyToUse=true
 
+	// Snapshot is ready - update copy trigger if necessary
+	err = vh.updateCopyTriggerAfterCloneOrSnap(ctx, src)
+	if err != nil {
+		return snap, err
+	}
+
 	logger.V(1).Info("temporary snapshot reconciled", "operation", op)
 	return snap, nil
 }
 
+// Checks if copy trigger annotations are on the srcPVC and updates accordingly
+// This should be called when clone/snapshot does not exist in order to determine
+// if we should wait before creation.
+//
+//nolint:funlen
+func (vh *VolumeHandler) waitForCopyTriggerBeforeCloneOrSnap(ctx context.Context, log logr.Logger,
+	srcPVC *corev1.PersistentVolumeClaim) (bool, error) {
+	if !utils.PVCUsesCopyTrigger(srcPVC) {
+		// No need to wait
+		return false, nil
+	}
+
+	logger := log.WithValues("useCopyTrigger", "true")
+	// Look for user supplied copy trigger
+	copyTrigger := utils.GetCopyTriggerValue(srcPVC)
+	latestCopyTrigger := utils.GetLatestCopyTriggerValue(srcPVC)
+	if copyTrigger == "" || copyTrigger == latestCopyTrigger {
+		// Need to block until the user updates the copyTrigger
+		wait := true
+
+		updated := utils.SetLatestCopyStatusWaitingForTrigger(srcPVC)
+		if updated {
+			err := vh.client.Update(ctx, srcPVC)
+			if err != nil {
+				return wait, err
+			}
+			vh.eventRecorder.Eventf(vh.owner, srcPVC, corev1.EventTypeNormal,
+				volsyncv1alpha1.EvRSrcPVCWaitingForCopyTrigger, volsyncv1alpha1.EvACreateSrcCopyUsingCopyTrigger,
+				"Waiting on copy trigger on src PVC %s before creating snapshot or clone",
+				utils.KindAndName(vh.client.Scheme(), srcPVC))
+		}
+
+		waitingSinceTimeStr := utils.GetLatestCopyTriggerWaitingSinceValue(srcPVC)
+		waitingSinceTime, err := time.Parse(time.RFC3339, waitingSinceTimeStr)
+		if err != nil {
+			logger.Error(err, "Waiting since pvc Annotation is not parseable, resetting ...")
+			utils.SetLatestCopyTriggerWaitingSinceValueNow(srcPVC)
+			err := vh.client.Update(ctx, srcPVC)
+			if err != nil {
+				return wait, err
+			}
+		}
+		if waitingSinceTime.Add(utils.CopyTriggerWaitTimeout).Before(time.Now()) {
+			copyTriggerTimeoutErr := &volsyncerrors.CopyTriggerTimeoutError{
+				SourcePVC: srcPVC.GetName(),
+			}
+			logger.Error(copyTriggerTimeoutErr, fmt.Sprintf("Timed out after %s", utils.CopyTriggerWaitTimeout.String()))
+			vh.eventRecorder.Eventf(vh.owner, srcPVC, corev1.EventTypeWarning,
+				volsyncv1alpha1.EvRSrcPVCTimeoutWaitingForCopyTrigger, volsyncv1alpha1.EvACreateSrcCopyUsingCopyTrigger,
+				"Timed out waiting on copy trigger on src PVC %s before creating snapshot or clone",
+				utils.KindAndName(vh.client.Scheme(), srcPVC))
+			return wait, &volsyncerrors.CopyTriggerTimeoutError{
+				SourcePVC: srcPVC.GetName(),
+			}
+		}
```

Comment:
Looking at this... If we have a timeout, we'll return an error, but we won't proceed w/ the replication. Then when they do finally provide the annotation, we'll proceed just like if there had been no timeout, correct?

Reply:
That's correct - the timeout is really just a point at which I put up this error in the latestMoverStatus as we've been waiting for the copy-trigger to be updated for a set period of time - so similar to where we start creating events for services not coming up on time.

Reply:
Looking at this again, I made an update to not use the word timeout in the event or the message (I just log at .Info now).  I think it is a bit misleading to say we've "Timed out" as we're still going to keep reconciling.